# Análise exploratória dos datasets

Este notebook realiza apenas a inspeção inicial da estrutura e da qualidade das fontes disponíveis.

## Preparação do ambiente

In [1]:
from pathlib import Path

import pandas as pd
from IPython.display import display

# Permite executar o notebook tanto pela raiz do projeto quanto pela pasta do notebook.
candidatos_raiz = [Path.cwd(), *Path.cwd().parents]
RAIZ_PROJETO = next(
    caminho for caminho in candidatos_raiz
    if (caminho / 'data' / 'raw').is_dir()
)
PASTA_DADOS = RAIZ_PROJETO / 'data' / 'raw'
PASTA_DADOS

WindowsPath('c:/Users/Paola/OneDrive/Documentos/Projetos/Pós IA Scientist/Tech Challenge/Fase 2/Indicador Criança Alfabetizada/literacy-data-pipeline/data/raw')

## 1. `avaliacao_alfabetizacao_municipio.csv`

### Importação e leitura do arquivo

In [2]:
caminho_avaliacao_municipio = PASTA_DADOS / 'avaliacao_alfabetizacao_municipio.csv'
avaliacao_municipio = pd.read_csv(caminho_avaliacao_municipio)

### Primeiros registros, dimensões e colunas

In [3]:
display(avaliacao_municipio.head())
print(f'Quantidade de linhas e colunas: {avaliacao_municipio.shape}')
avaliacao_municipio.columns.tolist()

,ano,id_municipio,serie,rede,taxa_alfabetizacao,media_portugues,proporcao_aluno_nivel_0,proporcao_aluno_nivel_1,proporcao_aluno_nivel_2,proporcao_aluno_nivel_3,proporcao_aluno_nivel_4,proporcao_aluno_nivel_5,proporcao_aluno_nivel_6,proporcao_aluno_nivel_7,proporcao_aluno_nivel_8
0,2023,1100031,2,3,69.10,767.8763,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2023,1100072,2,3,58.20,747.8918,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2023,1100189,2,5,69.73,762.4062,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2023,1101609,2,3,50.70,745.6802,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2023,1101807,2,3,55.69,752.3724,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Quantidade de linhas e colunas: (23995, 15)


['ano',
 'id_municipio',
 'serie',
 'rede',
 'taxa_alfabetizacao',
 'media_portugues',
 'proporcao_aluno_nivel_0',
 'proporcao_aluno_nivel_1',
 'proporcao_aluno_nivel_2',
 'proporcao_aluno_nivel_3',
 'proporcao_aluno_nivel_4',
 'proporcao_aluno_nivel_5',
 'proporcao_aluno_nivel_6',
 'proporcao_aluno_nivel_7',
 'proporcao_aluno_nivel_8']

### Tipos, registros e valores não nulos

In [4]:
avaliacao_municipio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23995 entries, 0 to 23994
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ano                      23995 non-null  int64  
 1   id_municipio             23995 non-null  int64  
 2   serie                    23995 non-null  int64  
 3   rede                     23995 non-null  int64  
 4   taxa_alfabetizacao       23995 non-null  float64
 5   media_portugues          23995 non-null  float64
 6   proporcao_aluno_nivel_0  12448 non-null  float64
 7   proporcao_aluno_nivel_1  12448 non-null  float64
 8   proporcao_aluno_nivel_2  12448 non-null  float64
 9   proporcao_aluno_nivel_3  12448 non-null  float64
 10  proporcao_aluno_nivel_4  12448 non-null  float64
 11  proporcao_aluno_nivel_5  12448 non-null  float64
 12  proporcao_aluno_nivel_6  12448 non-null  float64
 13  proporcao_aluno_nivel_7  12448 non-null  float64
 14  proporcao_aluno_nivel_

### Estatísticas descritivas

In [5]:
display(avaliacao_municipio.describe(include='all'))

,ano,id_municipio,serie,rede,taxa_alfabetizacao,media_portugues,proporcao_aluno_nivel_0,proporcao_aluno_nivel_1,proporcao_aluno_nivel_2,proporcao_aluno_nivel_3,proporcao_aluno_nivel_4,proporcao_aluno_nivel_5,proporcao_aluno_nivel_6,proporcao_aluno_nivel_7,proporcao_aluno_nivel_8
count,23995.000000,2.399500e+04,23995.0,23995.000000,23995.000000,23995.000000,12448.000000,12448.000000,12448.000000,12448.000000,12448.000000,12448.000000,12448.000000,12448.000000,12448.000000
mean,2023.518775,3.266174e+06,2.0,3.729444,61.444545,751.375708,2.201809,4.308785,7.733355,11.274998,18.268491,24.216200,18.629434,8.907910,4.458915
std,0.499658,9.824721e+05,0.0,1.206273,19.743521,23.232837,2.689885,4.296197,6.313496,7.261559,7.824829,9.487374,9.228701,7.583738,9.088799
min,2023.000000,1.100015e+06,2.0,0.000000,2.120000,673.298300,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2023.000000,2.514453e+06,2.0,3.000000,47.170000,736.035750,0.000000,0.830000,2.847500,5.760000,13.330000,17.777500,11.760000,3.360000,0.000000
50%,2024.000000,3.143302e+06,2.0,3.000000,62.230000,750.654200,1.520000,3.320000,6.580000,10.440000,18.370000,23.650000,17.725000,7.140000,1.900000
75%,2024.000000,4.127007e+06,2.0,5.000000,76.550000,764.302900,3.280000,6.492500,11.390000,16.052500,23.110000,30.522500,24.380000,12.180000,4.680000
max,2024.000000,5.300108e+06,2.0,5.000000,100.000000,868.455400,30.000000,50.000000,47.060000,59.730000,64.710000,70.000000,68.750000,61.540000,89.710000


### Valores nulos, duplicidades e valores únicos

In [6]:
display(avaliacao_municipio.isnull().sum().rename('quantidade_nulos').to_frame())
print(f'Registros completamente duplicados: {avaliacao_municipio.duplicated().sum()}')
display(avaliacao_municipio.nunique().rename('quantidade_valores_unicos').to_frame())

,quantidade_nulos
ano,0
id_municipio,0
serie,0
rede,0
taxa_alfabetizacao,0
media_portugues,0
proporcao_aluno_nivel_0,11547
proporcao_aluno_nivel_1,11547
proporcao_aluno_nivel_2,11547
proporcao_aluno_nivel_3,11547


Registros completamente duplicados: 0


,quantidade_valores_unicos
ano,2
id_municipio,5550
serie,1
rede,4
taxa_alfabetizacao,6161
media_portugues,13321
proporcao_aluno_nivel_0,943
proporcao_aluno_nivel_1,1374
proporcao_aluno_nivel_2,1922
proporcao_aluno_nivel_3,2289


### Observações

- **Possível chave primária:** Combinação `ano` + `id_municipio` + `rede`, sem repetições na fonte atual. A coluna `serie` possui apenas o valor 2 e pode integrar a definição formal do grão caso outras séries sejam incorporadas futuramente.
- **Possíveis chaves de relacionamento:** `id_municipio` com fontes municipais; `ano` e `rede` para compatibilizar período e dependência administrativa.
- **Colunas relevantes:** `ano`, `id_municipio`, `serie`, `rede`, `taxa_alfabetizacao`, `media_portugues` e as nove colunas `proporcao_aluno_nivel_0` a `proporcao_aluno_nivel_8`.
- **Problemas encontrados:** Não há registros completamente duplicados. As nove colunas de proporção possuem 11.547 nulos cada, correspondentes a todos os registros de 2023; em 2024 elas estão preenchidas. `serie` é constante (2) e os códigos de `rede` precisam de documentação de domínio.
- **Tratamentos necessários na Silver:** Validar a unicidade da chave composta em cada carga; documentar os nulos de 2023 como ausência estrutural até confirmação da regra de negócio; validar domínios de `rede`, `serie` e `id_municipio`; preservar a granularidade e os valores originais com rastreabilidade.

## 2. `avaliacao_alfabetizacao_uf.csv`

### Importação e leitura do arquivo

In [7]:
caminho_avaliacao_uf = PASTA_DADOS / 'avaliacao_alfabetizacao_uf.csv'
avaliacao_uf = pd.read_csv(caminho_avaliacao_uf)

### Primeiros registros, dimensões e colunas

In [8]:
display(avaliacao_uf.head())
print(f'Quantidade de linhas e colunas: {avaliacao_uf.shape}')
avaliacao_uf.columns.tolist()

,ano,sigla_uf,serie,rede,taxa_alfabetizacao,media_portugues,proporcao_aluno_nivel_0,proporcao_aluno_nivel_1,proporcao_aluno_nivel_2,proporcao_aluno_nivel_3,proporcao_aluno_nivel_4,proporcao_aluno_nivel_5,proporcao_aluno_nivel_6,proporcao_aluno_nivel_7,proporcao_aluno_nivel_8
0,2023,AM,2,3,49.20,733.6637,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2023,PB,2,2,55.23,744.8152,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2023,PR,2,5,73.12,757.2146,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2023,AP,2,3,41.87,732.7858,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2023,PE,2,5,58.95,747.4522,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Quantidade de linhas e colunas: (145, 15)


['ano',
 'sigla_uf',
 'serie',
 'rede',
 'taxa_alfabetizacao',
 'media_portugues',
 'proporcao_aluno_nivel_0',
 'proporcao_aluno_nivel_1',
 'proporcao_aluno_nivel_2',
 'proporcao_aluno_nivel_3',
 'proporcao_aluno_nivel_4',
 'proporcao_aluno_nivel_5',
 'proporcao_aluno_nivel_6',
 'proporcao_aluno_nivel_7',
 'proporcao_aluno_nivel_8']

### Tipos, registros e valores não nulos

In [9]:
avaliacao_uf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145 entries, 0 to 144
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ano                      145 non-null    int64  
 1   sigla_uf                 145 non-null    object 
 2   serie                    145 non-null    int64  
 3   rede                     145 non-null    int64  
 4   taxa_alfabetizacao       145 non-null    float64
 5   media_portugues          145 non-null    float64
 6   proporcao_aluno_nivel_0  75 non-null     float64
 7   proporcao_aluno_nivel_1  75 non-null     float64
 8   proporcao_aluno_nivel_2  75 non-null     float64
 9   proporcao_aluno_nivel_3  75 non-null     float64
 10  proporcao_aluno_nivel_4  75 non-null     float64
 11  proporcao_aluno_nivel_5  75 non-null     float64
 12  proporcao_aluno_nivel_6  75 non-null     float64
 13  proporcao_aluno_nivel_7  75 non-null     float64
 14  proporcao_aluno_nivel_8  7

### Estatísticas descritivas

In [10]:
display(avaliacao_uf.describe(include='all'))

,ano,sigla_uf,serie,rede,taxa_alfabetizacao,media_portugues,proporcao_aluno_nivel_0,proporcao_aluno_nivel_1,proporcao_aluno_nivel_2,proporcao_aluno_nivel_3,proporcao_aluno_nivel_4,proporcao_aluno_nivel_5,proporcao_aluno_nivel_6,proporcao_aluno_nivel_7,proporcao_aluno_nivel_8
count,145.000000,145,145.0,145.000000,145.000000,145.000000,75.000000,75.000000,75.000000,75.000000,75.000000,75.000000,75.00000,75.000000,75.000000
unique,NaN,25,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,BA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,2023.517241,NaN,2.0,3.337931,56.374000,745.336640,3.049600,5.491200,9.025733,12.011467,19.128267,23.647067,16.74960,7.291467,3.604667
std,0.501435,NaN,0.0,1.275957,13.190042,15.785106,1.657242,2.406019,3.914428,4.597241,5.134635,6.428589,5.42381,4.652399,5.966252
min,2023.000000,NaN,2.0,0.000000,30.570000,712.562000,0.000000,1.410000,2.440000,1.520000,6.110000,10.140000,8.24000,2.100000,0.000000
25%,2023.000000,NaN,2.0,2.000000,47.450000,734.254700,1.750000,3.965000,6.000000,8.065000,16.345000,18.780000,13.17500,3.660000,0.925000
50%,2024.000000,NaN,2.0,3.000000,55.870000,744.815200,2.890000,5.420000,8.920000,12.310000,18.870000,24.230000,15.80000,5.770000,1.940000
75%,2024.000000,NaN,2.0,5.000000,63.550000,753.300000,3.860000,7.095000,10.995000,15.510000,22.440000,28.915000,19.84000,10.015000,3.655000


### Valores nulos, duplicidades e valores únicos

In [11]:
display(avaliacao_uf.isnull().sum().rename('quantidade_nulos').to_frame())
print(f'Registros completamente duplicados: {avaliacao_uf.duplicated().sum()}')
display(avaliacao_uf.nunique().rename('quantidade_valores_unicos').to_frame())

,quantidade_nulos
ano,0
sigla_uf,0
serie,0
rede,0
taxa_alfabetizacao,0
media_portugues,0
proporcao_aluno_nivel_0,70
proporcao_aluno_nivel_1,70
proporcao_aluno_nivel_2,70
proporcao_aluno_nivel_3,70


Registros completamente duplicados: 0


,quantidade_valores_unicos
ano,2
sigla_uf,25
serie,1
rede,4
taxa_alfabetizacao,139
media_portugues,140
proporcao_aluno_nivel_0,60
proporcao_aluno_nivel_1,65
proporcao_aluno_nivel_2,68
proporcao_aluno_nivel_3,65


### Observações

- **Possível chave primária:** Combinação `ano` + `sigla_uf` + `rede`, sem repetições na fonte atual. A coluna `serie`, constante em 2, pode integrar a definição formal do grão se houver expansão para outras séries.
- **Possíveis chaves de relacionamento:** `sigla_uf` com fontes estaduais; `ano` e `rede` para compatibilizar período e dependência administrativa.
- **Colunas relevantes:** `ano`, `sigla_uf`, `serie`, `rede`, `taxa_alfabetizacao`, `media_portugues` e as nove colunas `proporcao_aluno_nivel_0` a `proporcao_aluno_nivel_8`.
- **Problemas encontrados:** Não há registros completamente duplicados. As nove colunas de proporção possuem 70 nulos cada, exatamente todos os registros de 2023; em 2024 elas estão preenchidas. A fonte contém 25 UFs distintas, exigindo verificação de cobertura, enquanto `serie` é constante (2).
- **Tratamentos necessários na Silver:** Validar a unicidade da chave composta e a cobertura das UFs; documentar os nulos de 2023 como ausência estrutural até confirmação; validar os domínios de `sigla_uf`, `rede` e `serie`; preservar a granularidade e os valores originais com rastreabilidade.

## 3. `meta_alfabetizacao_brasil.csv`

### Importação e leitura do arquivo

In [27]:
caminho_meta_brasil = PASTA_DADOS / 'meta_alfabetizacao_brasil.csv'
meta_brasil = pd.read_csv(caminho_meta_brasil)

### Primeiros registros, dimensões e colunas

In [28]:
display(meta_brasil.head())
print(f'Quantidade de linhas e colunas: {meta_brasil.shape}')
meta_brasil.columns.tolist()

,ano,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,percentual_participacao
0,2025,Pública,66.0,60.0,64.00,67.00,71.00,74.00,77.00,80,88.00
1,2024,Pública,59.2,59.9,63.77,67.47,70.97,74.23,77.24,80,87.37
2,2023,Pública,55.9,59.9,63.77,67.47,70.97,74.23,77.24,80,86.00


Quantidade de linhas e colunas: (3, 11)


['ano',
 'rede',
 'taxa_alfabetizacao',
 'meta_alfabetizacao_2024',
 'meta_alfabetizacao_2025',
 'meta_alfabetizacao_2026',
 'meta_alfabetizacao_2027',
 'meta_alfabetizacao_2028',
 'meta_alfabetizacao_2029',
 'meta_alfabetizacao_2030',
 'percentual_participacao']

### Tipos, registros e valores não nulos

In [14]:
meta_brasil.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ano                      3 non-null      int64  
 1   rede                     3 non-null      object 
 2   taxa_alfabetizacao       3 non-null      float64
 3   meta_alfabetizacao_2024  3 non-null      float64
 4   meta_alfabetizacao_2025  3 non-null      float64
 5   meta_alfabetizacao_2026  3 non-null      float64
 6   meta_alfabetizacao_2027  3 non-null      float64
 7   meta_alfabetizacao_2028  3 non-null      float64
 8   meta_alfabetizacao_2029  3 non-null      float64
 9   meta_alfabetizacao_2030  3 non-null      int64  
 10  percentual_participacao  3 non-null      float64
dtypes: float64(8), int64(2), object(1)
memory usage: 396.0+ bytes


### Estatísticas descritivas

In [15]:
display(meta_brasil.describe(include='all'))

,ano,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,percentual_participacao
count,3.0,3,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.0,3.000000
unique,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,Pública,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,2024.0,NaN,60.366667,59.933333,63.846667,67.313333,70.980000,74.153333,77.160000,80.0,87.123333
std,1.0,NaN,5.150081,0.057735,0.132791,0.271355,0.017321,0.132791,0.138564,0.0,1.022562
min,2023.0,NaN,55.900000,59.900000,63.770000,67.000000,70.970000,74.000000,77.000000,80.0,86.000000
25%,2023.5,NaN,57.550000,59.900000,63.770000,67.235000,70.970000,74.115000,77.120000,80.0,86.685000
50%,2024.0,NaN,59.200000,59.900000,63.770000,67.470000,70.970000,74.230000,77.240000,80.0,87.370000
75%,2024.5,NaN,62.600000,59.950000,63.885000,67.470000,70.985000,74.230000,77.240000,80.0,87.685000


### Valores nulos, duplicidades e valores únicos

In [16]:
display(meta_brasil.isnull().sum().rename('quantidade_nulos').to_frame())
print(f'Registros completamente duplicados: {meta_brasil.duplicated().sum()}')
display(meta_brasil.nunique().rename('quantidade_valores_unicos').to_frame())

,quantidade_nulos
ano,0
rede,0
taxa_alfabetizacao,0
meta_alfabetizacao_2024,0
meta_alfabetizacao_2025,0
meta_alfabetizacao_2026,0
meta_alfabetizacao_2027,0
meta_alfabetizacao_2028,0
meta_alfabetizacao_2029,0
meta_alfabetizacao_2030,0


Registros completamente duplicados: 0


,quantidade_valores_unicos
ano,3
rede,1
taxa_alfabetizacao,3
meta_alfabetizacao_2024,2
meta_alfabetizacao_2025,2
meta_alfabetizacao_2026,2
meta_alfabetizacao_2027,2
meta_alfabetizacao_2028,2
meta_alfabetizacao_2029,2
meta_alfabetizacao_2030,1


### Observações

- **Possível chave primária:** Combinação `ano` + `rede`, sem repetições na fonte atual. Como `rede` possui somente o valor `Pública`, `ano` também identifica unicamente os três registros observados, mas a chave composta explicita melhor o grão.
- **Possíveis chaves de relacionamento:** `ano` e `rede` com as fontes de metas por UF e município e com agregações nacionais de avaliação, observando a compatibilidade dos domínios de rede.
- **Colunas relevantes:** `ano`, `rede`, `taxa_alfabetizacao`, `meta_alfabetizacao_2024` a `meta_alfabetizacao_2030` e `percentual_participacao`.
- **Problemas encontrados:** Não há valores nulos nem registros completamente duplicados. A base possui apenas três registros (2023 a 2025), `rede` é constante e `meta_alfabetizacao_2030` também é constante em 80; essas características devem ser confirmadas como regras da fonte, não tratadas automaticamente como erro.
- **Tratamentos necessários na Silver:** Validar a unicidade por `ano` e `rede`; padronizar e documentar o domínio textual de `rede`; validar faixas percentuais e manter os anos das metas claramente identificados, sem imputar ou recalcular valores.

## 4. `meta_alfabetizacao_municipio.csv`

### Importação e leitura do arquivo

In [29]:
caminho_meta_municipio = PASTA_DADOS / 'meta_alfabetizacao_municipio.csv'
meta_municipio = pd.read_csv(caminho_meta_municipio)

### Primeiros registros, dimensões e colunas

In [30]:
display(meta_municipio.head())
print(f'Quantidade de linhas e colunas: {meta_municipio.shape}')
meta_municipio.columns.tolist()

,ano,id_municipio,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,nivel_alfabetizacao,percentual_participacao
0,2023,4301750,Municipal,NaN,NaN,14.05,23.65,37.00,52.68,67.85,80,NaN,NaN
1,2024,4301750,Municipal,4.40,NaN,14.05,23.65,37.00,52.68,67.85,80,0.0,92.59
2,2024,2406908,Municipal,42.86,7.94,14.05,23.65,37.00,52.68,67.85,80,1.0,84.00
3,2023,2406908,Municipal,4.40,7.94,14.05,23.65,37.00,52.68,67.85,80,0.0,82.14
4,2023,1718501,Municipal,4.60,8.25,14.48,24.16,37.49,53.03,68.00,80,0.0,95.65


Quantidade de linhas e colunas: (10704, 13)


['ano',
 'id_municipio',
 'rede',
 'taxa_alfabetizacao',
 'meta_alfabetizacao_2024',
 'meta_alfabetizacao_2025',
 'meta_alfabetizacao_2026',
 'meta_alfabetizacao_2027',
 'meta_alfabetizacao_2028',
 'meta_alfabetizacao_2029',
 'meta_alfabetizacao_2030',
 'nivel_alfabetizacao',
 'percentual_participacao']

### Tipos, registros e valores não nulos

In [31]:
meta_municipio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10704 entries, 0 to 10703
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ano                      10704 non-null  int64  
 1   id_municipio             10704 non-null  int64  
 2   rede                     10704 non-null  object 
 3   taxa_alfabetizacao       10584 non-null  float64
 4   meta_alfabetizacao_2024  10464 non-null  float64
 5   meta_alfabetizacao_2025  10704 non-null  float64
 6   meta_alfabetizacao_2026  10704 non-null  float64
 7   meta_alfabetizacao_2027  10704 non-null  float64
 8   meta_alfabetizacao_2028  10704 non-null  float64
 9   meta_alfabetizacao_2029  10704 non-null  float64
 10  meta_alfabetizacao_2030  10704 non-null  int64  
 11  nivel_alfabetizacao      10584 non-null  float64
 12  percentual_participacao  10584 non-null  float64
dtypes: float64(9), int64(3), object(1)
memory usage: 1.1+ MB


### Estatísticas descritivas

In [32]:
display(meta_municipio.describe(include='all'))

,ano,id_municipio,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,nivel_alfabetizacao,percentual_participacao
count,10704.000000,1.070400e+04,10704,10584.000000,10464.000000,10704.000000,10704.000000,10704.000000,10704.000000,10704.000000,10704.0,10584.000000,10584.000000
unique,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,NaN,Municipal,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,NaN,10704,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,2023.500000,3.246581e+06,NaN,61.774719,62.151745,65.245433,68.557849,71.771986,74.794895,77.552195,80.0,2.666667,90.427783
std,0.500023,9.844084e+05,NaN,19.527540,15.097461,12.632025,9.855499,7.060485,4.400614,2.018264,0.0,1.706356,6.270600
min,2023.000000,1.100015e+06,NaN,4.400000,7.940000,14.050000,23.650000,37.000000,52.680000,67.850000,80.0,0.000000,70.000000
25%,2023.000000,2.511177e+06,NaN,47.600000,52.170000,57.107500,62.397500,67.397500,72.037500,76.250000,80.0,1.000000,86.670000
50%,2023.500000,3.145030e+06,NaN,62.445000,64.095000,66.940000,69.875000,72.665000,75.285000,77.730000,80.0,3.000000,91.205000
75%,2024.000000,4.115866e+06,NaN,76.742500,75.885000,76.510000,77.240000,77.950000,78.650000,79.330000,80.0,4.000000,95.060000


### Valores nulos, duplicidades e valores únicos

In [33]:
display(meta_municipio.isnull().sum().rename('quantidade_nulos').to_frame())
print(f'Registros completamente duplicados: {meta_municipio.duplicated().sum()}')
display(meta_municipio.nunique().rename('quantidade_valores_unicos').to_frame())

,quantidade_nulos
ano,0
id_municipio,0
rede,0
taxa_alfabetizacao,120
meta_alfabetizacao_2024,240
meta_alfabetizacao_2025,0
meta_alfabetizacao_2026,0
meta_alfabetizacao_2027,0
meta_alfabetizacao_2028,0
meta_alfabetizacao_2029,0


Registros completamente duplicados: 0


,quantidade_valores_unicos
ano,2
id_municipio,5352
rede,1
taxa_alfabetizacao,3876
meta_alfabetizacao_2024,2745
meta_alfabetizacao_2025,2501
meta_alfabetizacao_2026,2284
meta_alfabetizacao_2027,1889
meta_alfabetizacao_2028,1387
meta_alfabetizacao_2029,755


### Observações

- **Possível chave primária:** Combinação `ano` + `id_municipio` + `rede`, sem repetições na fonte atual. Como `rede` é constante em `Municipal`, `ano` + `id_municipio` também é único nos dados observados, mas incluir `rede` explicita o grão.
- **Possíveis chaves de relacionamento:** `id_municipio` com a avaliação municipal e dimensões territoriais; `ano` e `rede` para compatibilizar período e dependência administrativa.
- **Colunas relevantes:** `ano`, `id_municipio`, `rede`, `taxa_alfabetizacao`, `meta_alfabetizacao_2024` a `meta_alfabetizacao_2030`, `nivel_alfabetizacao` e `percentual_participacao`.
- **Problemas encontrados:** Não há registros completamente duplicados. Existem 120 nulos em `taxa_alfabetizacao`, `nivel_alfabetizacao` e `percentual_participacao`, todos em 2023; `meta_alfabetizacao_2024` possui 240 nulos, divididos igualmente entre 2023 e 2024. `rede` e `meta_alfabetizacao_2030` são constantes, e o nível foi inferido como `float64` por causa dos nulos.
- **Tratamentos necessários na Silver:** Validar a chave composta e a cobertura dos municípios; investigar e documentar a regra dos nulos antes de qualquer imputação; validar domínios e faixas de percentuais e níveis; avaliar o tipo semântico de `nivel_alfabetizacao`; preservar os valores brutos e registrar regras de qualidade.

## 5. `meta_alfabetizacao_uf.csv`

### Importação e leitura do arquivo

In [34]:
caminho_meta_uf = PASTA_DADOS / 'meta_alfabetizacao_uf.csv'
meta_uf = pd.read_csv(caminho_meta_uf)

### Primeiros registros, dimensões e colunas

In [35]:
display(meta_uf.head())
print(f'Quantidade de linhas e colunas: {meta_uf.shape}')
meta_uf.columns.tolist()

,ano,sigla_uf,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,percentual_participacao
0,2024,RR,Pública,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2023,RR,Pública,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2023,AC,Pública,NaN,NaN,56.9,62.2,67.3,72.0,76.2,80.0,NaN
3,2024,AC,Pública,51.38,NaN,56.9,62.2,67.3,72.0,76.2,80.0,80.87
4,2023,DF,Pública,NaN,NaN,63.1,67.0,70.6,74.0,77.1,80.0,NaN


Quantidade de linhas e colunas: (54, 12)


['ano',
 'sigla_uf',
 'rede',
 'taxa_alfabetizacao',
 'meta_alfabetizacao_2024',
 'meta_alfabetizacao_2025',
 'meta_alfabetizacao_2026',
 'meta_alfabetizacao_2027',
 'meta_alfabetizacao_2028',
 'meta_alfabetizacao_2029',
 'meta_alfabetizacao_2030',
 'percentual_participacao']

### Tipos, registros e valores não nulos

In [36]:
meta_uf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ano                      54 non-null     int64  
 1   sigla_uf                 54 non-null     object 
 2   rede                     54 non-null     object 
 3   taxa_alfabetizacao       50 non-null     float64
 4   meta_alfabetizacao_2024  48 non-null     float64
 5   meta_alfabetizacao_2025  52 non-null     float64
 6   meta_alfabetizacao_2026  52 non-null     float64
 7   meta_alfabetizacao_2027  52 non-null     float64
 8   meta_alfabetizacao_2028  52 non-null     float64
 9   meta_alfabetizacao_2029  52 non-null     float64
 10  meta_alfabetizacao_2030  52 non-null     float64
 11  percentual_participacao  50 non-null     float64
dtypes: float64(9), int64(1), object(2)
memory usage: 5.2+ KB


### Estatísticas descritivas

In [37]:
display(meta_uf.describe(include='all'))

,ano,sigla_uf,rede,taxa_alfabetizacao,meta_alfabetizacao_2024,meta_alfabetizacao_2025,meta_alfabetizacao_2026,meta_alfabetizacao_2027,meta_alfabetizacao_2028,meta_alfabetizacao_2029,meta_alfabetizacao_2030,percentual_participacao
count,54.000000,54,54,50.000000,48.000000,52.000000,52.000000,52.000000,52.000000,52.000000,52.0,50.000000
unique,NaN,27,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,RR,Pública,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,2,54,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,2023.500000,NaN,NaN,55.531000,58.233333,62.188462,66.253846,70.126923,73.734615,77.038462,80.0,86.899000
std,0.504695,NaN,NaN,11.975824,10.021410,7.885620,6.059071,4.301533,2.659197,1.220439,0.0,5.636714
min,2023.000000,NaN,NaN,31.300000,38.300000,45.900000,53.600000,61.200000,68.300000,74.600000,80.0,70.050000
25%,2023.000000,NaN,NaN,48.250000,52.025000,56.900000,62.200000,67.300000,72.000000,76.200000,80.0,83.142500
50%,2023.500000,NaN,NaN,55.560000,56.900000,61.400000,65.650000,69.650000,73.450000,76.900000,80.0,88.310000
75%,2024.000000,NaN,NaN,61.860000,64.925000,67.400000,70.200000,72.900000,75.400000,77.800000,80.0,90.015000


### Valores nulos, duplicidades e valores únicos

In [38]:
display(meta_uf.isnull().sum().rename('quantidade_nulos').to_frame())
print(f'Registros completamente duplicados: {meta_uf.duplicated().sum()}')
display(meta_uf.nunique().rename('quantidade_valores_unicos').to_frame())

,quantidade_nulos
ano,0
sigla_uf,0
rede,0
taxa_alfabetizacao,4
meta_alfabetizacao_2024,6
meta_alfabetizacao_2025,2
meta_alfabetizacao_2026,2
meta_alfabetizacao_2027,2
meta_alfabetizacao_2028,2
meta_alfabetizacao_2029,2


Registros completamente duplicados: 0


,quantidade_valores_unicos
ano,2
sigla_uf,27
rede,1
taxa_alfabetizacao,50
meta_alfabetizacao_2024,24
meta_alfabetizacao_2025,26
meta_alfabetizacao_2026,24
meta_alfabetizacao_2027,24
meta_alfabetizacao_2028,23
meta_alfabetizacao_2029,21


### Observações

- **Possível chave primária:** Combinação `ano` + `sigla_uf` + `rede`, sem repetições na fonte atual. Como `rede` é constante em `Pública`, `ano` + `sigla_uf` também é único nos dados observados, mas a chave composta explicita o grão.
- **Possíveis chaves de relacionamento:** `sigla_uf` com a avaliação por UF e dimensões territoriais; `ano` e `rede` para compatibilizar período e dependência administrativa.
- **Colunas relevantes:** `ano`, `sigla_uf`, `rede`, `taxa_alfabetizacao`, `meta_alfabetizacao_2024` a `meta_alfabetizacao_2030` e `percentual_participacao`.
- **Problemas encontrados:** Não há registros completamente duplicados. Há 4 nulos em `taxa_alfabetizacao` e `percentual_participacao`, 6 em `meta_alfabetizacao_2024` e 2 em cada meta de 2025 a 2030. Os nulos ocorrem nos dois anos. `rede` é constante e `meta_alfabetizacao_2030` é 80 em todos os 52 registros preenchidos.
- **Tratamentos necessários na Silver:** Validar a chave composta e a cobertura das 27 UFs; investigar os padrões de ausência antes de qualquer imputação; padronizar e validar `sigla_uf` e `rede`; validar faixas percentuais e preservar os nulos com indicação de qualidade quando a fonte não informar o valor.

## Conclusões da Análise Exploratória

### Chaves e granularidade

- **avaliacao_alfabetizacao_municipio:** granularidade por município, ano e rede. Possível chave composta: `ano + id_municipio + rede`. A série é sempre 2 na fonte analisada.
- **avaliacao_alfabetizacao_uf:** granularidade por UF, ano e rede. Possível chave composta: `ano + sigla_uf + rede`. A série também é sempre 2.
- **meta_alfabetizacao_brasil:** granularidade nacional por ano e rede. Possível chave: `ano + rede`.
- **meta_alfabetizacao_municipio:** granularidade por município, ano e rede. Possível chave composta: `ano + id_municipio + rede`.
- **meta_alfabetizacao_uf:** granularidade por UF, ano e rede. Possível chave composta: `ano + sigla_uf + rede`.

As chaves foram consideradas possíveis chaves primárias porque não apresentaram repetições nos dados analisados. A unicidade deverá ser validada novamente durante o processamento da camada Silver.

### Relacionamentos identificados

Os principais relacionamentos identificados são:

- `avaliacao_alfabetizacao_municipio` ↔ `meta_alfabetizacao_municipio`
  - Chave territorial: `id_municipio`
  - Também devem ser considerados `ano` e a compatibilidade do domínio de `rede`.

- `avaliacao_alfabetizacao_uf` ↔ `meta_alfabetizacao_uf`
  - Chave territorial: `sigla_uf`
  - Também devem ser considerados `ano` e a compatibilidade do domínio de `rede`.

- `meta_alfabetizacao_brasil` possui granularidade nacional e poderá ser utilizada como referência para comparações agregadas, sem relacionamento direto por município ou UF.

### Problemas e características identificados

- Não foram encontrados registros completamente duplicados nas cinco fontes analisadas.
- Na avaliação municipal, as colunas `proporcao_aluno_nivel_0` até `proporcao_aluno_nivel_8` possuem valores nulos nos registros de 2023 e estão preenchidas em 2024.
- Na avaliação por UF ocorre comportamento semelhante nas colunas de proporção por nível.
- A avaliação por UF contém 25 UFs distintas, enquanto a base de metas por UF contém 27, indicando diferença de cobertura que deverá ser validada antes das integrações.
- A base de metas municipais possui valores ausentes em `taxa_alfabetizacao`, `nivel_alfabetizacao`, `percentual_participacao` e em metas de determinados períodos.
- `nivel_alfabetizacao` foi interpretado como `float64` devido à existência de valores nulos e deverá ter seu tipo semântico avaliado.
- A base de metas por UF também possui valores ausentes em indicadores e metas.
- Alguns campos são constantes nas fontes analisadas, como `serie = 2` nas avaliações e `meta_alfabetizacao_2030 = 80` nas metas preenchidas. Esses valores serão preservados, pois não há evidência na exploração de que representem erros.

### Decisões para a camada Silver

Na camada Silver serão realizadas as seguintes operações:

- Validar a unicidade das chaves compostas.
- Padronizar nomes e tipos das colunas.
- Validar códigos de município e siglas das UFs.
- Validar os domínios de `rede` e `serie`.
- Validar as faixas dos indicadores percentuais.
- Tratar duplicidades somente caso sejam identificadas durante novas cargas.
- Preservar valores nulos quando representarem ausência de informação na fonte, evitando imputações sem justificativa.
- Registrar diferenças de cobertura territorial como regra de qualidade.
- Avaliar e ajustar o tipo de `nivel_alfabetizacao`.
- Integrar as fontes somente após a validação das chaves e da compatibilidade de granularidade.
- Preservar a rastreabilidade dos dados provenientes da camada Bronze.